tambola/static/game.js

```diff
@@ -1,105 +1,72 @@
 const winPatterns = [
   {
-    'name': 'Full House',
-    'validator': fullHouseValidator
-  }, 
+    name: 'Full House',
+    validator: (t) =>
+      !t.data.flat().some((cell) => cell !== '-' && cell !== 'x'),
+  },
   {
-    'name': 'Top Row',
-    'validator': topRowValidator
+    name: 'Top Row',
+    validator: (t) => !t.data[0].some((cell) => cell !== '-' && cell !== 'x'),
   },
   {
-    'name': 'Early 5',
-    'validator': earlyValidator
-  }
+    name: 'Early 5',
+    validator: (t) => t.data.flat().filter((cell) => cell === 'x').length >= 5,
+  },
 ];
 
-function fullHouseValidator(ticket) {
-  for(let i = 0; i < ticket.data.length; i++) {
-    for(let j = 0; j < ticket.data[i].length; j++) {
-      if(ticket.data[i][j] !== "-" && ticket.data[i][j] !== "x") {
-        return false;
-      }
-    }
-  }
-  return true;
-}
-
-function topRowValidator(ticket) {
-  for(let j = 0; j < ticket.data[0].length; j++) {
-    if(ticket.data[0][j] !== "-" && ticket.data[0][j] !== "x") {
-      return false;
-    }
-  }
-  return true;
-}
-
-function earlyValidator(ticket) {
-  let xCount = 0;
-  for(let i = 0; i < ticket.data.length; i++) {
-    for(let j = 0; j < ticket.data[i].length; j++) {
-      if(ticket.data[i][j] === "x") {
-        xCount++;
-      }
-    }
-  }
-  return xCount >= 5;
-}
-
 showWinners();
 
 async function showWinners() {
   const tickets = await getTickets();
   const markedNums = await getMarkedNums();
-  const markedTickets = tickets.map(ticket => markTicket(ticket, markedNums));
+  const markedTickets = tickets.map((ticket) => markTicket(ticket, markedNums));
 
   const rootDiv = document.getElementById('win_patterns');
-  rootDiv.innerHTML = "";
-  for(let i = 0; i < winPatterns.length; i++) {
-    const pattern = winPatterns[i];
-    rootDiv.innerHTML += ('<div>');
-    rootDiv.innerHTML += ('<h3>' + pattern.name + '</h3>');
-    const wonThisPattern = markedTickets.filter(pattern.validator);
-    rootDiv.innerHTML += ('<ul>');
-    for(let j = 0; j < wonThisPattern.length; j++) {
-      rootDiv.innerHTML += ('<li>' + wonThisPattern[j].name + '</li>');
-    }
-    rootDiv.innerHTML += ('</ul>');
-    rootDiv.innerHTML += ('</div>');
-  }
+  rootDiv.innerHTML = '';
+  winPatterns.forEach((pattern) => {
+    rootDiv.innerHTML += `
+    <div>
+      <h3>${pattern.name}</h3>
+      <ul>
+      ${markedTickets
```

Comment:
Wait, you can do this in string interpolation? guess I should brush up my Javascript